index.js

```diff
@@ -348,140 +348,104 @@ alexa.app = function(name) {
-            var eventHandlerObject = self.audioPlayerEventHandlers[event];
-            if (typeof eventHandlerObject != "undefined" && typeof eventHandlerObject["function"] == "function") {
-              var eventHandlerResult = eventHandlerObject["function"](request, response, callbackHandler);
-              if (eventHandlerObject && eventHandlerObject.then) {
-                Promise.resolve(eventHandlerResult).asCallback(callbackHandler);
-              } else if (false !== eventHandlerResult) {
-                callbackHandler();
-              } else {
-                console.trace("NOTE: using `return false` for async audio player requests is deprecated and will not work after the next major version");
-              }
-            } else {
-              response.send();
-            }
+        // propagate successful response if it's already been resolved
+        return response.response;
+      });
+    };
+
+    return promiseChain.then(function () {
+      // Call to `.pre` can also throw, so we wrap it in a promise here to
+      // propagate errors to the error handler
+      var prePromise = Promise.resolve();
+      if (typeof self.pre == "function") {
+        prePromise = Promise.resolve(self.pre(request, response, requestType));
+      }
+      return prePromise;
+    }).then(function () {
+      if (!response.resolved) {
+        if ("IntentRequest" === requestType) {
+          var intent = request_json.request.intent.name;
+          if (typeof self.intents[intent] != "undefined" && typeof self.intents[intent]["function"] == "function") {
+            return Promise.resolve(self.intents[intent]["function"](request, response));
+          } else {
+            throw "NO_INTENT_FOUND";
+          }
+        } else if ("LaunchRequest" === requestType) {
+          if (typeof self.launchFunc == "function") {
+            return Promise.resolve(self.launchFunc(request, response));
           } else {
-            throw "INVALID_REQUEST_TYPE";
+            throw "NO_LAUNCH_FUNCTION";
+          }
+        } else if ("SessionEndedRequest" === requestType) {
+          if (typeof self.sessionEndedFunc == "function") {
+            return Promise.resolve(self.sessionEndedFunc(request, response));
+          }
+        } else if (request.isAudioPlayer()) {
+          var event = requestType.slice(12);
+          var eventHandlerObject = self.audioPlayerEventHandlers[event];
+          if (typeof eventHandlerObject != "undefined" && typeof eventHandlerObject["function"] == "function") {
+            return Promise.resolve(eventHandlerObject["function"](request, response));
           }
+        } else {
+          throw "INVALID_REQUEST_TYPE";
+        }
+      }
+    })
+    .then(function () {
+      return response.send();
+    })
+    .catch(function(e) {
+      if (typeof self.error == "function") {
+        self.error(e, request, response);
+      } else if (typeof e == "string" && self.messages[e]) {
```

Comment:
I think the default behavior should be to resolve by calling `response.send()`here as the user error handler, `app.error()` can still explicitly fail if desired.